pkg/storegateway/bucket_index_postings.go

```diff
@@ -393,15 +396,33 @@ func (s worstCaseFetchedDataStrategy) selectPostings(groups []postingGroup) (sel
 		atLeastOneIntersectingSelected bool
 		maxSelectedSize                = maxSelectedSeriesCount * tsdb.EstimatedSeriesP99Size
 	)
+	selected = groups
 	for i, g := range groups {
 		postingListSize := int64(float64(g.totalSize) * s.postingListActualSizeFactor)
 		if atLeastOneIntersectingSelected && selectedSize+postingListSize > maxSelectedSize {
-			return groups[:i], groups[i:]
+			selected = groups[:i]
+			omitted = groups[i:]
+			break
 		}
 		selectedSize += postingListSize
 		atLeastOneIntersectingSelected = atLeastOneIntersectingSelected || !g.isSubtract
 	}
-	return groups, nil
+
+	// We want to include the __name__ group because excluding it is more likely to make
+	// makes the more sparse in the index. Selecting more sparse series results in more API calls
+	// to the object store. This is because series are first sorted by their __name__ label
+	// (assuming there are no labels starting with uppercase letters).
+	for i, g := range omitted {
+		if len(g.keys) > 0 && g.keys[0].Name == labels.MetricName {
+			// Since the underlying slice for selected and omitted in the same, we need to swap the group so that
+			// we don't overwrite the first group when we append to selected.
+			omitted[0], omitted[i] = omitted[i], omitted[0]
+			omitted = omitted[1:]
+			selected = append(selected, g)
```

Comment:
This could also be a `selected = selected[0: len(selected)+1]` correct? I'm wondering if doing so may be more clear than using `append()`.

Reply:
yeah, sure 👍 
```suggestion
			selected = selected[0: len(selected)+1]
```